src/System/IO/Extra.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_HADDOCK hide #-}
+
+module System.IO.Extra
+    (
+    -- * I/O application-specific helpers
+    -- ** Read
+      hGetBytes
+    , hGetSomeMnemonic
+    , hGetXPrv
+    , hGetXPub
+    , hGetXP__
+
+    -- ** Write
+    , hPutBold
+    , hPutBytes
+
+    -- * I/O Helpers
+    , failWith
+    , withSGR
+    ) where
+
+import Prelude
+
+import Cardano.Address.Derivation
+    ( XPrv, XPub, xprvFromBytes, xpubFromBytes )
+import Cardano.Mnemonic
+    ( MkSomeMnemonicError (..), SomeMnemonic, mkSomeMnemonic )
+import Control.Applicative
+    ( (<|>) )
+import Control.Exception
+    ( bracket )
+import Data.ByteArray.Encoding
+    ( Base (..), convertFromBase, convertToBase )
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Base58
+    ( bitcoinAlphabet, decodeBase58, encodeBase58 )
+import Data.Either.Extra
+    ( eitherToMaybe )
+import Options.Applicative.Encoding
+    ( AbstractEncoding (..), Encoding )
+import System.Console.ANSI
+    ( Color (..)
+    , ColorIntensity (..)
+    , ConsoleIntensity (..)
+    , ConsoleLayer (..)
+    , SGR (..)
+    , hSetSGR
+    )
+import System.Exit
+    ( exitFailure )
+import System.IO
+    ( Handle, hIsTerminalDevice, stderr )
+
+import qualified Codec.Binary.Bech32 as Bech32
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+
+--
+-- I/O Read
+--
+
+-- | Read some bytes from the console, and decode them if the encoding is recognized.
+hGetBytes :: Handle -> IO ByteString
+hGetBytes h = do
+    raw <- B8.hGetContents h
+    case (tryBech32 raw <|> tryBase16 raw <|> tryBase58 raw) of
```

Comment:
No, I'm actually concerned here
```haskell
λ> tryAny "11"
"\DC1"
λ> tryBase58 "11"
Just "\NUL\NUL"
```
This function does not have access to the expected length of the decoded bytes, so it get's even trickier.

Reply:
It doesn't actually matter much does it? The bytestrings manipulated by this program are at least 32 bytes (so ~64 characters minimum hex encoded).

There's no way a 32 bytestrings can be both base58 and base16 decoded.

Reply:
- What if you enter a `base58` key that is also valid hex-bytes (but not key), then wouldn't `hGetBytes` choose hex, and the key-from-bytes function fail, instead of successfully decoding a base58 key?
- I suspect error messages will be _much_ worse
- There is also (I believe) ambiguity of xpub (64) vs xprv (96). If someone in the future added something of size 128: we could silently miss that there is no way do distinguish between 128byte base58 and 64 byte hex.

Reply:
Some kind of top-level roundtrip tests (like in previous implementation) should catch these. (At least 1, and 3)

Reply:
> What if you enter a base58 key that is also valid hex-bytes (but not key), then wouldn't hGetBytes choose hex, and the key-from-bytes function fail, instead of successfully decoding a base58 key?

What do you mean?

> I suspect error messages will be much worse

At this stage, they will be worse indeed. 

> There is also (I believe) ambiguity of xpub (64) vs xprv (96). If someone in the future added something of size 128: we could silently miss that there is no way do distinguish between 128byte base58 and 64 byte hex.

I disagree here. The alphabet is not the same! The probability that a string of 50 characters encoded in Base58 would only contain characters from the base16 alphabet is 15/58**50 ~ 1e-30 !! And I am not even sure that this makes for a valid encoding! Really, I don't think there's any issue.

Reply:
> What do you mean?

I was imagining "that a string in Base58 would [...] only contain characters from the base16"

> The probability that a string of

Ok, fair point.

Then just: You wrote "especially if base16 is tried last", but doesn't seem to be done here

Also ran in to something strange
```
# A root key
$ cardano-wallet-shelley key root -- art forum devote street sure rather head chuckle guard poverty release quote oak craft enemy
b8f2bece9bdfe2b0282f5bad705562ac996efb6af96b648f4445ec44f47ad95c10e3d72f26ed075422a36ed8585c745a0e1150bcceba2357d058636991f38a3791e248de509c070d812ab2fda57860ac876bc489192c1ef4ce253c197ee219a4

# Can derive child from recovery phrase using cardano-address 
$ echo "art forum devote street sure rather head chuckle guard poverty release quote oak craft enemy" | cardano-address key from-recovery-phrase icarus --base16 | cardano-address key child 0
xprv13zahs5rgyl8a60xkaxkpucvlz48fl08qtq2cukatde5nr7r6m9w0xnjkvvaq233mf045q7wn2h83t65f85l95au049kn3yk35vwla2w72hpcpm2v0e0ezkd6kh3p823qn2n3zmle200en5t3whxf5zj375kw8aes

# ... but can not go from the root key directly?
$ echo "b8f2bece9bdfe2b0282f5bad705562ac996efb6af96b648f4445ec44f47ad95c10e3d72f26ed075422a36ed8585c745a0e1150bcceba2357d058636991f38a3791e248de509c070d812ab2fda57860ac876bc489192c1ef4ce253c197ee219a4" | cardano-address key child "0"
Couldn't convert bytes into neither extended public or private keys.

# Works with old
$ echo "b8f2bece9bdfe2b0282f5bad705562ac996efb6af96b648f4445ec44f47ad95c10e3d72f26ed075422a36ed8585c745a0e1150bcceba2357d058636991f38a3791e248de509c070d812ab2fda57860ac876bc489192c1ef4ce253c197ee219a4" | cardano-wallet-shelley key child --path 0
88bb78506827cfdd3cd6e9ac1e619f154e9fbce058158e5bab6e6931f87ad95cf34e56633a05463b4beb4079d355cf15ea893d3e5a778fa96d3892d1a31dfea9de55c380ed4c7e5f9159bab5e213aa209aa7116ff953df99d17175cc9a0a51f5

```

Reply:
```console
# ... but can not go from the root key directly?
$ echo "b8f2bece9bdfe2b0282f5bad705562ac996efb6af96b648f4445ec44f47ad95c10e3d72f26ed075422a36ed8585c745a0e1150bcceba2357d058636991f38a3791e248de509c070d812ab2fda57860ac876bc489192c1ef4ce253c197ee219a4" | cardano-address key child "0"
Couldn't convert bytes into neither extended public or private keys.
```

Indeed! `echo` adds a trailing `\n` that needs to be trimmed. Now fixed.